src/dune_engine/fs_memo.ml

```diff
@@ -55,27 +68,42 @@ let watch_path dune_file_watcher path =
     | Ok () ->
       ())
 
-let watch_path_using_ref path =
+let watch_path_using_ref kind path =
   match !t_ref with
   | Initialized { dune_file_watcher = None } -> ()
-  | Initialized { dune_file_watcher = Some watcher } -> watch_path watcher path
+  | Initialized { dune_file_watcher = Some watcher } ->
+    watch_path watcher (kind, path)
   | Uninitialized paths_to_watch ->
-    t_ref := Uninitialized (path :: paths_to_watch)
+    t_ref := Uninitialized ((kind, path) :: paths_to_watch)
 
 (* Files and directories have non-overlapping sets of paths, so we can track
    them using the same memoization table. *)
 let memo =
+  let module Input = struct
+    type t = watch * Path.t
+
+    let to_dyn = Dyn.pair dyn_of_watch Path.to_dyn
+
+    let hash = Tuple.T2.hash Poly.hash Path.hash
+
+    let equal (k1, p1) (k2, p2) =
+      match (k1, k2) with
+      | Parent, Parent
+      | This, This ->
+        Path.equal p1 p2
+      | _, _ -> false
+  end in
```

Comment:
I think squashing this in one table makes it harder to think about what's going on, because paths are actually semantically different. Paths that are watched via parent are files, and paths that are watched directly are directories. Using two tables also seems simpler since you don't need to wrap things into `Input`.

Reply:
> Paths that are watched via parent are files, and paths that are watched directly are directories.

That's not exactly the case I think. If we want to stat a path that is a directory, we want to stat its parent even if its already a directory. Do you agree?

Reply:
Yes, I think you are right: 

* The `Via_parent` table can contain both files and directories.
* The `Directly` table can contain only directories.

Uh, the `path_stat` case is super confusing.